// vue-data-ui/dist/vue-ui-kpi-CgRnp_cJ.js
import { defineAsyncComponent as A, computed as x, useSlots as z, onMounted as C, watch as y, ref as k, createElementBlock as i, openBlock as s, normalizeStyle as u, normalizeClass as r, createElementVNode as $, renderSlot as n, createTextVNode as h, toDisplayString as b, createVNode as w, unref as v, Fragment as B } from "vue";
import { u as L } from "./useNestedProp-Bw1KcAmj.js";
import { u as M, i as E, f as I } from "./index-BLtEpj8j.js";
const G = {
  __name: "vue-ui-kpi",
  props: {
    config: {
      type: Object,
      default() {
        return {};
      }
    },
    dataset: {
      type: Number,
      default: 0
    }
  },
  setup(o) {
    const F = A(() => import("./vue-ui-digits-DOcvXxeY.js")), { vue_ui_kpi: N } = M(), l = o, e = x({
      get: () => c(),
      set: (a) => a
    }), D = z();
    C(() => {
      D["chart-background"] && console.warn("VueUiKpi does not support the #chart-background slot.");
    });
    function c() {
      return L({
        userConfig: l.config,
        defaultConfig: N
      });
    }
    y(() => l.config, (a) => {
      e.value = c(), d();
    }, { deep: !0 });
    const S = k((typeof l.dataset == "number", l.dataset)), t = k(e.value.useAnimation ? e.value.animationValueStart : S.value), m = (a) => {
      const f = e.value.animationFrames, p = Math.abs(a - t.value) / f;
      function g() {
        t.value < a ? t.value = Math.min(t.value + p, a) : t.value > a && (t.value = Math.max(t.value - p, a)), t.value !== a && requestAnimationFrame(g);
      }
      g();
    };
    C(() => {
      d();
    });
    function d() {
      e.value.useAnimation ? (t.value = e.value.animationValueStart, m(l.dataset)) : t.value = l.dataset;
    }
    return y(() => l.dataset, (a) => {
      e.value.useAnimation ? m(a) : t.value = a;
    }), (a, f) => (s(), i("div", {
      class: r(`vue-ui-kpi ${e.value.layoutClass}`),
      style: u(`background:${e.value.backgroundColor}; ${e.value.layoutCss}`)
    }, [
      $("div", {
        class: r(`vue-ui-kpi-title ${e.value.titleClass}`),
        style: u(`font-family: ${e.value.fontFamily}; font-size:${e.value.titleFontSize}px; color:${e.value.titleColor}; font-weight:${e.value.titleBold ? "bold" : "normal"}; ${e.value.titleCss}`)
      }, [
        n(a.$slots, "title", { comment: o.dataset }),
        h(" " + b(e.value.title), 1)
      ], 6),
      n(a.$slots, "comment-before", { comment: o.dataset }),
      $("div", {
        class: r(`vue-ui-kpi-value ${e.value.valueClass}`),
        style: u(`font-family: ${e.value.fontFamily}; font-size:${e.value.valueFontSize}px; color:${e.value.valueColor}; font-weight:${e.value.valueBold ? "bold" : "normal"}; ${e.value.valueCss}`)
      }, [
        n(a.$slots, "value", { comment: o.dataset }),
        e.value.analogDigits.show ? (s(), i("div", {
          key: 0,
          style: u({ height: e.value.analogDigits.height + "px" })
        }, [
          w(v(F), {
            dataset: Number(t.value.toFixed(e.value.valueRounding)),
            config: {
              backgroundColor: e.value.backgroundColor,
              digits: {
                color: e.value.analogDigits.color,
                skeletonColor: e.value.analogDigits.skeletonColor
              }
            }
          }, null, 8, ["dataset", "config"])
        ], 4)) : (s(), i(B, { key: 1 }, [
          h(b(v(E)(
            e.value.formatter,
            t.value,
            v(I)({
              p: e.value.prefix,
              v: t.value,
              s: e.value.suffix,
              r: e.value.valueRounding
            })
          )), 1)
        ], 64))
      ], 6),
      n(a.$slots, "comment-after", { comment: o.dataset })
    ], 6));
  }
};
export {
  G as default
};
